### ISSUE TITLE
[Bug] AccuracyAwareQuantization fails while trying to compute statistics

### ISSUE THREAD

**Issue by stentll** [openvinotoolkit/openvino#7841]
##### System information (version)

- OpenVINO => 2021.4.689
- Operating System / Platform => WSL2 on Windows 64 Bit
- Problem classification: Model Quantization
- Framework: OpenVINO (using IR model files)

##### Detailed description
I am trying to quantize a model using AccuracyAwareQuantization but after some success I stumbled upon a hurdle I cannot handle.

I tried following the documentation on this at [Python API docs page](https://docs.openvinotoolkit.org/latest/pot_compression_api_README.html#overview) and got to a point where I could create a pipeline using `create pipeline` function: 
`compression.pipline.initializer.create_pipeline(algo_config, engine)`

First of all, the docs have a typo in 'pipline' and secondly this gave me an error saying `AttributeError: module 'compression.pipeline' has no attribute 'initializer'`

After walking around in the source code I found that perhaps the most up to date substitution for this function would be using the following class: `compression.AccuracyAwareQuantization(algorithms, engine)`. It had a run() method which seemed to fit my expectations so I continued on with this. This is where the documentation became very hard to find (I eventually based half of my code looking at OpenVINO source code and figuring out what is expected of my DataLoader, Metric etc. In addition I looked at [this notebook](https://github.com/openvinotoolkit/openvino_notebooks/blob/main/notebooks/105-language-quantize-bert/105-language-quantize-bert.ipynb) which again was not easy to find from any docs links.

I continued with using AccuracyAwareQuantization and cannot get past this error:

```
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
/tmp/ipykernel_742/746459721.py in <module>
      2 logging.basicConfig(level=logging.DEBUG)
      3 # Step 6: Execute the pipeline.
----> 4 compressed_model = pipeline.run(model)

/opt/intel/openvino_2021/deployment_tools/tools/post_training_optimization_toolkit/compression/algorithms/quantization/accuracy_aware/algorithm.py in run(self, model)
    166         # quantize model
    167         quantized_model, metrics_accuracy_drop, quantized_metrics_per_sample = \
--> 168             self._quantize_and_evaluate(deepcopy(model),
    169                                         self._quantize_model,
    170                                         print_progress=print_progress)

/opt/intel/openvino_2021/deployment_tools/tools/post_training_optimization_toolkit/compression/algorithms/quantization/accuracy_aware/algorithm.py in _quantize_and_evaluate(self, model, quantization_algo, print_progress)
    470                     for metric_name, params in self._metrics_config.items()}
    471 
--> 472         quantized_model = quantization_algo(model)
    473         logger.info('Start compressed model inference')
    474         quantized_metrics, quantized_metrics_per_sample = \

/opt/intel/openvino_2021/deployment_tools/tools/post_training_optimization_toolkit/compression/algorithms/quantization/accuracy_aware/algorithm.py in _quantize_model(self, model)
    458     def _quantize_model(self, model):
    459         logger.info('Start quantization')
--> 460         return self._quantization_algo.run(model)
    461 
    462     def _search_optimal_parameters(self, model):

/opt/intel/openvino_2021/deployment_tools/tools/post_training_optimization_toolkit/compression/algorithms/quantization/default/algorithm.py in run(self, model)
     86             print(self.algorithms[0].algo_collector)
     87             logger.info('Start computing statistics for algorithm : {}'.format(self.algorithms[0].name))
---> 88             self.algorithms[0].algo_collector.compute_statistics(model)
     89             logger.info('Computing statistics finished')
     90 

AttributeError: 'NoneType' object has no attribute 'compute_statistics'
```

I printed out `self.algorithms` at this point and got:
`[<compression.algorithms.quantization.channel_alignment.algorithm.ActivationChannelAlignment object at 0x7fac6d6006d0>, <compression.algorithms.quantization.minmax.algorithm.MinMaxQuantization object at 0x7fac6fa7e2b0>, <compression.algorithms.quantization.fast_bias_correction.algorithm.FastBiasCorrection object at 0x7fac6d82e9a0>]`

##### Steps to reproduce

```python
# Step 1: Load the model.
model_config = {
        'model_name': 'PPUnet',
        'model': MODEL_XML,
        'weights': MODEL_BIN
}

model = compression.graph.model_utils.load_model(model_config)

# Step 2: Initialize the data loader. 
# Custom DataLoader for segmentation
dataset_config = {
    'df': df, 
    'img_cols': ['img0_path', 'img11_path'], 
    'msk_cols': ['gt_path'], 
    'tile_size': 512,
    'mode': 'train'
} # dictionary with the dataset parameters 
data_loader = CustomDataLoader(**dataset_config)

# Step 3 (Optional. Required for AccuracyAwareQuantization): Initialize the metric.
metric = F1Metric() #F1 score for segmentation

 # Step 4: Initialize the engine for metric calculation and statistics collection.
engine_config = Dict({"device": "CPU"}) # dictionary with the engine parameters
engine = compression.engines.ie_engine.IEEngine(engine_config, data_loader, metric)

# Step 5: Initialize and run pipeline
# Dict here is means EasyDict because the config was otherwise crashing OpenVINO because in source code OpenVINO often fetches config params like an attribute

algorithms = Dict(
    {
        "target_device": "ANY",
        "metrics": [
            {
                "type": None,
                "name": "F1"
            }
        ],
        "ignored": {
        },

        "preset": "mixed", # A preset is a collection of optimization algorithm parameters
                           # that will specify to the algorithm to improve which metric
                           # the algorithm needs to concentrate. Each optimization algorithm
                           # supports [performance, mixed, accuracy] presets
        "optimization_subset_size": 300, # Size of subset to calculate activations statistics that can be used
                                 # for quantization parameters calculation
        "shuffle_data": False, # Shuffle data before selecting the subset to calculate activation
                               # statistics. An optional parameter, the default value is false
        "seed": 0, # Seed for data shuffle. An optional parameter, the default value is 0

        "weights": {
            "bits": 8, # Number of quantization bits
            "mode": "symmetric", # Quantization mode
            "granularity": "perchannel", # Granularity: a scale for each output channel
            "level_low": -127, # Low quantization level
            "level_high": 127, # High quantization level

            # /* Parameters specify how to calculate the minimum and maximum of quantization range */
            "range_estimator": {
                "max": {
                    "type": "quantile",
                    "outlier_prob": 0.0001
                }
            }
        },

        # /* Quantization parameters for activations */
        "activations": {
            "bits": 8, # Number of quantization bits
            "mode": "symmetric", # Quantization mode
            "granularity": "pertensor", # Granularity: one scale for output tensor

            # /* Parameters specify how to calculate the minimum and maximum of quantization range */
            "range_estimator": {
                "preset": "quantile",
            }
        }
    }
)


pipeline = compression.AccuracyAwareQuantization(algorithms, engine)
compressed_model = pipeline.run(model)
```

My used metric class looks like this:

```python
class F1Metric(compression.api.Metric):
    
    def __init__(self):
        super().__init__()
        self._name = "F1"
        self._scores = []
    
    @property
    def value(self):
        """Returns F1 metric value for the last model output."""
        return {self._name: self._scores[-1]}
    
    @property
    def avg_value(self):
        """Returns F1 metric value for all model outputs."""
        return {self._name: np.ravel(self._scores).mean()}
    
    def update(self, output, target):
        """
        Updates prediction matches.

        :param output: model output
        :param target: annotations
        """
        if len(output) > 1:
            raise Exception(
                "The F1 metric cannot be calculated " "for a model with multiple outputs"
            )
        output = output[0]#[0]
        target = target[0]
        output = imgutils.threshold(output, 0.5)
        assert output.shape == target.shape

        # using logical_ functions here because they are fast
        tp = np.sum(1 * np.logical_and(output, target))
        fp = np.sum(1 * np.logical_and(output, np.logical_not(target)))
        fn = np.sum(1 * np.logical_and(np.logical_not(output), target))
        f1: float = 2 * tp / (2 * tp + fp + fn)
        
        self._scores.append([f1])

    def reset(self):
        """
        Resets collected matches
        """
        self._matches = []
    
    def get_attributes(self):
        """
        Returns a dictionary of metric attributes {metric_name: {attribute_name: value}}.
        Required attributes: 'direction': 'higher-better' or 'higher-worse'
                             'type': metric type
        """
        return {self._name: {"direction": "higher-better", "type": "F1"}}   
```

(Sidenote: it is weird that DataLoader output arrays need an index and when DataLoader outputs arrays then those get converted to List[np.ndarray] inside OpenVINO which caused much confusion and debugging.)

Unfortunately I cannot include model nor input data but I think the issue is more of a proper documentation issue and maybe even with the code itself. 

The input images are with shape (1,2,512,512) and segmentation masks (1,3,512,512). I was using batch_size of 1.

##### Issue submission checklist

 - [x] I report the issue, it's not a question
   
 - [x] I checked the problem with documentation, FAQ, open issues, Stack Overflow, etc and have not found solution
   
 - [ ] There is reproducer code and related data files: images, videos, models, etc.

**Comment by Iffa-Meah** [comment]
In that case, have you ever consider trying out the[ POT CLI](https://docs.openvinotoolkit.org/latest/pot_configs_examples_README.html) instead to see whether it work with your model?
Generally, once you had set up the POT, you'll need to prepare a dataset (probably cut it if necessary), use the correct config file and you'll get the Quantized INT8 model.

Note that generally, MYRIAD/NCS2 requires model in FP16 format. However Quantizing FP16 will change it into INT8 format.
Currently, [INT8 in OpenVINO for VPU is not supported](https://docs.openvinotoolkit.org/latest/openvino_docs_IE_DG_supported_plugins_Supported_Devices.html#supported_model_formats). (Note that VPU,MYRIAD,NCS2 are referring to the same thing)

**Comment by stentll** [comment]
I took Python API to work with because the Python API docs said two things:

> The Python* POT API provides simple interfaces for implementing:
> data loading and pre-processing on an arbitrary dataset,
> custom accuracy metrics,

And what I am dealing with are a non-standard dataset and a non-standard accuracy metric. I did not find an easy way to use POT CLI to work with my dataset. My model is a fully convolutional segmentation model with two-channel input and 3-channel output and my metric is F1 score. In addition, my dataset has the following specifics:

- The two channels in input are glued together into one image during runtime and are stored as individual files on disk beforehand
- The input images are larger than the model input and are cropped during runtime
- The target mask is a single channel image that gets converted into a 3-channel target dynamically during runtime


As I understood from your response I should prepare the dataset beforehand to fit the POT CLI expectations but how can I set up a custom accuracy metric with POT CLI?

**Comment by jgespino** [comment]
Hi @stentll 

> First of all, the docs have a typo in 'pipline' and secondly this gave me an error saying AttributeError: module 'compression.pipeline' has no attribute 'initializer'

We have fixed the typo in PR #9337, thank you for reporting!

I have confirmed with the development team that the CLI will not work with your model as the model contains multiple inputs. We recommend looking at other [samples](https://github.com/openvinotoolkit/openvino/tree/master/tools/pot/openvino/tools/pot/api/samples) for the Post-training Optimization Tool API.

Regards,
Jesus
